examples/fluids/src/setupstg_shur14.c

```diff
@@ -0,0 +1,134 @@
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+/// @file
+/// Implementation of the Synthetic Turbulence Generation (STG) algorithm
+/// presented in Shur et al. 2014
+
+#include <stdlib.h>
+#include <math.h>
+#include "../qfunctions/stg_shur14.h"
+
+#ifndef M_PI
+#define M_PI    3.14159265358979323846
+#endif
+
+#ifndef NOMINMAX
+
+#ifndef max
+#define max(a,b)            (((a) > (b)) ? (a) : (b))
+#endif
+
+#ifndef min
+#define min(a,b)            (((a) < (b)) ? (a) : (b))
+#endif
+
+#endif  /* NOMINMAX */
+
+/*
+ * @brief Perform Cholesky decomposition on array of symmetric 3x3 matrices
+ *
+ * This assumes the input matrices are in order [11,22,33,12,13,23]. This
+ * format is also used for the output.
+ *
+ * @param[in] nprofs Number of matrices in Rij
+ * @param[in] Rij Array of the symmetric matrices [6,nprofs]
+ * @param[out] Cij Array of the Cholesky Decomposition matrices, [6,nprofs]
+ */
+static inline int CalcCholeskyDecomp(int nprofs, const CeedScalar Rij[6][nprofs], CeedScalar Cij[6][nprofs]){
+
+  CeedPragmaSIMD
+  for(int i=0; i<nprofs; i++){
+    Cij[0][i] = sqrt(Rij[0][i]);
+    Cij[3][i] = Rij[3][i] / Cij[0][i];
+    Cij[1][i] = sqrt(Rij[1][i] - pow(Cij[3][i], 2) );
+    Cij[4][i] = Rij[4][i] / Cij[0][i];
+    Cij[5][i] = (Rij[5][i] - Cij[3][i]*Cij[4][i]) / Cij[1][i];
+    Cij[2][i] = sqrt(Rij[2][i] - pow(Cij[4][i], 2) - pow(Cij[5][i], 2));
+  }
+return 1;
+}
+
+
+int SetupSTG_Rand(STGShur14Context stg_ctx){
+
+  //TODO will probably want to have the paths to the STGRand.dat and
+  // STGInflow.dat as inputs for this function.
+
+
+  //TODO Read STGRand.dat to get nmodes
+  int nmodes = 2;
+  //TODO Read STGInflow.dat to get nprof
+  int nprofs = 5;
+
+  int total_num_scalars = nmodes*(3 + 3 + 1) + nprofs*(1 + 3 + 6 + 1 + 1) ;
+  stg_ctx = malloc(sizeof(STGShur14Context) + total_num_scalars*sizeof(CeedScalar) );
+
+  stg_ctx->nmodes = nmodes;
+  stg_ctx->nprofs = nprofs;
+
+  stg_ctx->offsets.sigma = 0;
+  stg_ctx->offsets.d       = nmodes*3;
+  stg_ctx->offsets.phi     = stg_ctx->offsets.d       + nmodes*3;
+  stg_ctx->offsets.kappa   = stg_ctx->offsets.phi     + nmodes;
+  stg_ctx->offsets.prof_dw = stg_ctx->offsets.kappa   + nmodes;
+  stg_ctx->offsets.ubar    = stg_ctx->offsets.prof_dw + nprofs;
+  stg_ctx->offsets.cij     = stg_ctx->offsets.ubar    + nprofs*3;
+  stg_ctx->offsets.eps     = stg_ctx->offsets.ubar    + nprofs*6;
+  stg_ctx->offsets.lt      = stg_ctx->offsets.eps     + nprofs;
```

Comment:
I think the size calculation is too small (misses one `nmodes`), but here's a different way to initialize that might be less error-prone.
```suggestion
  {
    STGShur14Context s;
    s.nmodes = nmodes;
    s.nprofs = nprofs;
    s.offsets.sigma = 0;
    s.offsets.d       = nmodes*3;
    s.offsets.phi     = stg_ctx->offsets.d       + nmodes*3;
    s.offsets.kappa   = stg_ctx->offsets.phi     + nmodes;
    s.offsets.prof_dw = stg_ctx->offsets.kappa   + nmodes;
    s.offsets.ubar    = stg_ctx->offsets.prof_dw + nprofs;
    s.offsets.cij     = stg_ctx->offsets.ubar    + nprofs*3;
    s.offsets.eps     = stg_ctx->offsets.ubar    + nprofs*6;
    s.offsets.lt      = stg_ctx->offsets.eps     + nprofs;
    int total_num_scalars = stg_ctx->offsets.lt + nprofs;
    stg_ctx = malloc(sizeof(*stg_ctx) + total_num_scalars*sizeof(stg_ctx->data[0]));
    *stg_ctx = s;
  }
```

Reply:
Yep, you're right. Thanks for the suggestion!